### ISSUE TITLE
Make the `AnalysisBase` compatible with live simulation streams

### ISSUE THREAD

**Issue by ljwoods2** [MDAnalysis/mdanalysis#4828]
## Is your feature request related to a problem? ##

Related to #4827: for MDAnalysis to function with live simulation streams, the `AnalysisBase` must be modified to account for a few quirks of live-streamed trajectories:
- Their length/end frame isn't known (so `len(u.trajectory)` and `u.trajectory.n_frames` will fail)
- They can't be sliced in any way except `u.trajectory[:]` and `u.trajectory[::n]` (where n is positive) so they are inherently incompatible with parallelization schemes

## Describe the solution you'd like ##

Like with parallelization, analysis classes that inherit from `AnalysisBase` should "opt-in" to stream usage through a class property similar to the `parallelizable` property (like `streamable` or similar). To opt-in, inherited classes must also avoid attempting to access the trajectory's length or slice it. The `rms.RMSF` class is already such a class.

The `AnalysisBase` must also have a `_streamed_run` method that never attempts to slice a trajectory or access its length. The [imdclient MDAKit](https://github.com/Becksteinlab/imdclient/tree/main/imdclient) ships such a method already (and monkey patches it in): https://github.com/Becksteinlab/imdclient/blob/main/imdclient/streamanalysis.py

## Describe alternatives you've considered ##

Don't allow live simulation streaming, or don't allow builtin analysis classes to be used with these streams

**Comment by orbeckst** [comment]
Can this issue be addressed independently from #4827 ?

**Comment by ljwoods2** [comment]
> Can this issue be addressed independently from #4827 ?

Fixing this relies on being able to check if a reader inherits from `StreamReaderBase` (and failing if a stream reader tries to use an analysis class that isn't streamable) which would be a part of #4827- so good point, they are not entirely independent. The rest of the changes would be independent.

**Comment by yuxuanzhuang** [comment]
Generally, I advocate for including essential (and lightweight) components from `IMDClient` into MDAnalysis. Here are a few thoughts and comments:

- Initialization of Results:
I’m uncertain how you plan to handle the initialization of results for (almost all) analyses (e.g., [example in RMS.py](https://github.com/MDAnalysis/mdanalysis/blob/4f143fdf02ddf9cdb43b0e69edc7cb41bb455556/package/MDAnalysis/analysis/rms.py#L677)), considering you don’t know the `n_frames` beforehand.

- Streamable Analyses:
Some analyses may or may not be streamable under certain conditions. For example, in the case of `RMSD`, I assume that if the reference is an `IMDReader` and `ref_frame` is not zero, it would prevent the analysis from being streamable.

- IMDReader Compatibility:
Instead of modifying `AnalysisBase` to be streamable, would it be feasible to make `IMDReader` “AnalysisBase-compatible”? That is if necessary changes for `IMDReader` itself can be made to make iteration over `IMDReader` behaves similarly to other Readers?

- Slicing and Stopping Iteration in IMDReader:
Expanding on that, I wonder if disallowing the `stop` parameter for slicing `IMDReader` is strictly necessary. I may be missing some details, but it seems theoretically possible to stop `IMDReader` iteration after processing n frames. If `IMDReader` could support slicing (`start=0, step, stop`), it might naturally enable setting `n_frames` based on the slice.

**Comment by ljwoods2** [comment]
@yuxuanzhuang Thank you for the in-depth look!

Overall, I agree with you that we could adapt the reader to be more compatible.

As for the initialization of results, my thinking was that we would use python lists to hold intermediate results, and when iteration stops and `_conclude` is called, the list is converted to a numpy array and made available exactly as before.

Maybe we allow slicing, but raise a `RuntimeError` if the start frame isn't reached or the end frame isn't reached? What do we pass as the end frame if no `stop` is defined? In this case, we still would need the AnalysisBase modifications to handle an unknown-length stream- though this could be avoided if we **require** the `stop` kwarg for stream readers. In this case, should we still raise an error if that frame isn't reached? Or should we give people their intermediate results?

That might be the best path forward- requiring people to pass in an end frame, defaulting to 0 for the start frame, and failing out if the end frame isn't reached, perhaps with some kind of option to save intermediate results to csv or something. That would also allow us to avoid adding complexity with parsing logic for all the different analysis class' frame-related kwargs like `ref_frame`, and we could initialize results arrays with the required end frame arg as the length.

Edit: Say something like `ref_frame` is a frame in the middle of the trajectory- what do we do with the frames up until the `ref_frame` which isn't yet available (can't just jump directly to it)?

**Comment by yuxuanzhuang** [comment]
> As for the initialization of results, my thinking was that we would use python lists to hold intermediate results, and when iteration stops and _conclude is called, the list is converted to a numpy array and made available exactly as before.

Many analyses also rely on array assignment within `single_frame`, so the list-appending approach wouldn’t work in these cases or unless you want to refactor the whole `AnalysisBase`. For example, see [this usage in MSD analysis](https://github.com/MDAnalysis/mdanalysis/blob/b8fe34b73c9df9330c1608229b2f8cddc6e275b4/package/MDAnalysis/analysis/msd.py#L381).

I don’t have a fully fleshed-out solution yet, but I wonder if it might be possible to implement a “block” analysis approach so it could be processed in chunks of a fixed number of frames, and then the results from each block could be concatenated during `_stream_run` at the end.

**Comment by orbeckst** [comment]
Good idea. 

I hope that the discussion on this issue leads to a feasible and general solution, similar to how the parallelization was refined through discussions. It‘s important for the future of MDAnalysis.

**Comment by hmacdope** [comment]
> I don’t have a fully fleshed-out solution yet, but I wonder if it might be possible to implement a “block” analysis approach so it could be processed in chunks of a fixed number of frames, and then the results from each block could be concatenated during `_stream_run` at the end.

This is a really clever idea IMO, if the trajectory could be processed in blocks guaranteeing that `size == blocksize` or  `size == mod(full_len, blocksize)` for the end of the stream , then each can call a full AnalysisBase run and then concatenate the resulting results objects. Is this what you were thinking @yuxuanzhuang? It is similar to the way we approached parallelisation IIRC.

**Comment by IAlibay** [comment]
I haven't had time to think this through fully yet, but one initial comment / thought based on @yuxuanzhuang's comment "Instead of modifying AnalysisBase to be streamable, would it be feasible to make IMDReader “AnalysisBase-compatible”?"

I'm not sure if it's in the same spirit, but I am concerned that we are following the path of "let's keep on adding things to AnalysisBase because we can".

It sounds to me like enabling analyses on streams is going to need a lot of analysis-specific work. Is this a fair assessment? If so, is it maybe worth consider if we need either a mixin or a separate subclass to make analyses stream compatible? (rather than adding more to the one base class)

**Comment by yuxuanzhuang** [comment]
> This is a really clever idea IMO, if the trajectory could be processed in blocks guaranteeing that `size == blocksize` or `size == mod(full_len, blocksize)` for the end of the stream , then each can call a full AnalysisBase run and then concatenate the resulting results objects. Is this what you were thinking @yuxuanzhuang? It is similar to the way we approached parallelisation IIRC.

Yes, exactly. With such, I think the existing `_get_aggregator` for parallel analysis should just be the right tool for concatenation so no extra function will be needed for that.

**Comment by PardhavMaradani** [comment]
I am here from PR #5184 that is attempting to add support to run `AnalysisBase`-based classes for the current frame of streamed trajectories (This is for a specific visualization use case in `MolecularNodes` as described in issue #5183 ). With that limited experience, here are some of my thoughts on this issue:

- We cannot allow moving the stream pointer during any analysis. If we did, we will not be able to support running other analysis classes on the same set of frames. So, whatever `block` or `batch` of frames needed for an analysis need to be setup at a level above and not within `AnalysisBase`
- The workaround suggested by @marinegor  for issue #5183 might work well here - create a temporary in-memory universe after reading a `block`/`batch` of frames and run the analysis classes on it. This should work with all of them in their current form and also supports parallelization. There would still be a need to support aggregation of the results from all these independent `run`'s, but the `ResultsGroup` already provides a great way to do that
- If the above is not user friendly, we could extend all the above ideas a bit. The imd workshop [Application 2](https://github.com/amruthesht/imd-workshop-2025/blob/main/workshop/workshop.ipynb) presents the idea of a `batch` (similar to a `block` mentioned by Yuxuan above) which has a certain number of frames. Currently, `IMDReader` holds just one frame in memory. Similar to how `buffer_size` can be configured when setting it up, we could potentially configure a `batch_size` and hold those many frames in-memory. For all practical purposes, a streamed trajectory could then behave like a regular trajectory within a batch (while allowing frame number continuity across batches). This would allow re-use of existing `AnalysisBase`-based classes, support parallelization within the batch, etc. We would need an explicit way to move to the next batch and also a way to aggregate results across `run`'s as before - maybe through a new independent `ResultsAggregator` or something like that because this can be useful for others too

**Comment by PardhavMaradani** [comment]
Re-initiating this discussion based on additional data points and experiences from getting `AnalysisBase`-based classes [working](https://mdadash.readthedocs.io/en/latest/batching.html#analysisbase-support) for streaming in [mdadash](https://github.com/MDAnalysis/mdadash):

Considering a solution purely within the context of `AnalysisBase` will likely make this very restrictive. Since a streamed trajectory can only be iterated forward and most engines [don't support](https://imdclient.readthedocs.io/en/latest/usage.html) multiple concurrent connections, running through an entire simulation for a single analysis is probably a no go (first bullet of the previous comment).

Instead, we could consider running multiple analyses together by introducing something like an `AnalysisGroup` or `AnalysisEnsemble` or something along those lines. Here is how this could work - using the name `AnalysisGroup` here for illustration.

First, from the user perspective:

- Users create different `AnalysisBase`-based class instances as they'd normally do
- Users pass a list of these instances to create an `AnalysisGroup` instance
  - There will be validations to ensure they all point to the same underlying trajectory instance, etc
- Instead of running individual `.run(...)`'s, a `.run(...)` of this `AnalysisGroup` instance is called for these cases
  - For streaming, either all or only the `step` param is currently supported for iteration
  - We will be able to support non-serial backends for streaming as well (details below)
- The `results` object is accessed from this `AnalysisGroup` instance (and not individual analysis instances)
  - This is an array of individual analysis result objects corresponding to the input analysis list
  - Invididual analysis results object would be exactly the same compared to an individual `.run(...)` results - including aggregations, etc

The `AnalysisGroup` internally takes care of iterating the trajectory, running the different analyses instances, aggregating the results etc as follows:

- It maintains a sliding window buffer of the last n frames
  - similar to the [BufferedTrajectory](https://imdclient.readthedocs.io/en/latest/usage.html) in `mdadash`
- The buffer / batch size will be a configurable param
- Once a batch of timesteps is ready, individual analyses instances are run using their `.run()` method
  - These instances see the `n_frames` as the batch / buffer size (ignoring corner cases here)
  - The trajectory iteration that happens as part of `run()` are returned the timesteps from the buffer (see [BufferedTrajectory](https://mdadash.readthedocs.io/en/latest/autosummary/mdadash.backend.kernel.core.html#mdadash.backend.kernel.core.BufferedTrajectory) of `mdadash` as an example)
  - Native parallelization built into `AnalysisBase` with a non-serial backend can be leveraged as is
    - Since analyses instances passed can have different levels of support for parallelization, either an implicit fallback or an explicit fallback order can be allowed for the `backend` param
    - `IMDReader`, which internally uses `imdclient` has a network socket, which makes the trajectory not serializable. `mdadash` patches `IMDReader` to remove `imdclient` from the serialization state which makes parallelization possible there (see [Parallelization in mdadash](https://mdadash.readthedocs.io/en/latest/parallelization.html)). This would be a safe change to do here too
- The results from each of the `.run()` for a given analysis instance are aggregated
  - This aggregation is no different to how it happens in the parallelization case across different computation group results
  - The `_conclude()` isn't needed for each of the individual `.run()`'s (see below on the `AnalysisBase` changes), and is only invoked in the `AnalysisGroup` once at the very end for each of the analyses instances 

The changes needed in `AnalysisBase` and existing Analysis classes are very minimal:

- `_prepare()` is already getting called multiple times - in the parallelization case from the `_compute` call of different computation groups
  - Most analysis classes already take this into account and have just setup the results object keys, etc
  - Some analyses have code in `_prepare()` that gets repeated because of this and from a quick review (needs an indepth look again) seems like it can be refactored safely
    - Note that some analyses will never work for streaming cases (like q1q2 contacts - which needs access to the last frame, or ones that require arbitrary seeks of the frames, etc as already documented)
- `_run()` needs to be re-entrant
  - Even without our streaming use case, I think it is normally expected that once an analysis instance is created, a user would invoke `.run(...)` with different params - say, with a different set of frames
  - This is the case already for most analyses, with some minor exceptions
    - dssp analysis for example, creates new keys in the results object during `_conclude()` and this fails during a subsequent run (in the parallelization case) because the aggregation isn't specified for these keys
    - These can be easily cleaned up
- Option to not invoke `_conclude()` at the end of a `_run()` - as an optional param to run
  - This is strictly not a must, but can avoid an unnecessary call
  - `AnalysisGroup` invokes this at the end for each analysis instance

The idea of an `AnalysisGroup` doesn't need to be limited to the streaming use cases. Given streaming is a more restrictive use case, anything that works for it will work for file based trajectories as well. Grouping multiple analyses (when possible) could help with reduced I/O, speed up due to single set of transformations for each timestep, etc. This seems like it could be generally useful. For streaming though, this approach seems to be one way to handle multiple analyses in the core library due to the inherent constraints.

Thoughts?

Thanks

**Comment by orbeckst** [comment]
FYI, there's PR #4017 for `AnalysisCollection` to address #3569. Did you have a look at those, too?

**Comment by PardhavMaradani** [comment]
> FYI, there's PR [#4017](https://github.com/MDAnalysis/mdanalysis/pull/4017) for `AnalysisCollection` to address [#3569](https://github.com/MDAnalysis/mdanalysis/issues/3569). Did you have a look at those, too?

Hi @orbeckst, thanks for the pointers. I hadn't seen these before. These were the exact use cases I had in mind when I mentioned this could be generally useful. The approach is obviously very different to the one here because the one here is built around the more restrictive streaming case first and hence more generic - it will address the use cases in the above issue / PR as well. The collection / group approach for analyses is a nice to have for regular cases, but a necessity for the streaming case. Thanks
